src/Veldrid/StagingMemoryPool.cs

```diff
@@ -1,114 +1,133 @@
-            byte[] array = _arrayPool.Rent(bytes.Length);
-            Array.Copy(bytes, array, bytes.Length);
-            return new StagingBlock(array, (uint)bytes.Length, this);
+            Rent((uint)bytes.Length, out StagingBlock block);
+            Marshal.Copy(bytes, 0, (IntPtr)block.Data, bytes.Length);
+            return block;
         }
 
-        public FixedStagingBlock GetFixedStagingBlock(uint sizeInBytes)
+        public StagingBlock GetStagingBlock(uint sizeInBytes)
         {
-            byte[] array = _arrayPool.Rent((int)sizeInBytes);
-            return new FixedStagingBlock(array, sizeInBytes, this);
+            Rent(sizeInBytes, out StagingBlock block);
+            Unsafe.InitBlock(block.Data, 0, sizeInBytes);
+            return block;
         }
 
-        public void Free(StagingBlock block)
+        public StagingBlock RetrieveById(uint id)
         {
-            bool clearArray = false;
-#if DEBUG
-            clearArray = true;
-#endif
-            _arrayPool.Return(block.Array, clearArray);
+            return _storage[(int)id];
         }
 
-        public void Free(FixedStagingBlock block)
+        private void Rent(uint size, out StagingBlock block)
         {
-            bool clearArray = false;
-#if DEBUG
-            clearArray = true;
-#endif
-            _arrayPool.Return(block.Array, clearArray);
+            lock (_lock)
+            {
+                SortedList<uint, uint> available = _availableBlocks;
+                IList<uint> indices = available.Values;
+                for (int i = 0; i < available.Count; i++)
+                {
+                    int index = (int)indices[i];
+                    StagingBlock current = _storage[index];
+                    if (current.Capacity >= size)
+                    {
+                        available.RemoveAt(i);
+                        current.SizeInBytes = size;
+                        block = current;
+                        _storage[index] = current;
+                        return;
+                    }
+                }
+
+                Allocate(size, out block);
+            }
         }
 
-        public void Free(byte[] array)
+        private void Allocate(uint sizeInBytes, out StagingBlock stagingBlock)
         {
-            bool clearArray = false;
-#if DEBUG
-            clearArray = true;
-#endif
-            _arrayPool.Return(array, clearArray);
+            IntPtr ptr = Marshal.AllocHGlobal((int)sizeInBytes);
+            uint id = (uint)_storage.Count;
+            stagingBlock = new StagingBlock(id, (void*)ptr, sizeInBytes, this);
+            _storage.Add(stagingBlock);
         }
-    }
 
-    internal unsafe struct StagingBlock
-    {
-        public readonly byte[] Array;
-        public readonly uint SizeInBytes;
-        public readonly StagingMemoryPool Pool;
+        public void Free(StagingBlock block)
+        {
+            lock (_lock)
+            {
+                if (block.Id < _storage.Count)
```

Comment:
It feels like this should be an assertion or something. What happens when a block with a nonsense ID is returned?

Reply:
>What happens when a block with a nonsense ID is returned
Nothing, and that's the whole point. I added this check so that calling ``Free`` wouldn't have any effect after the pool is disposed (and therefore _storage.Count is 0).

Reply:
Understood. It might be worth adding an assertion that the pool has been disposed if we hit this case, just to make it clear what the expectation is. Or, at least a comment of some sort. At first glance it's not clear what the check is for.